Firestore/core/src/firebase/firestore/local/local_store.cc

```diff
@@ -426,53 +456,57 @@ QueryData LocalStore::AllocateQuery(Query query) {
   // Sanity check to ensure that even when resuming a query it's not currently
   // active.
   TargetId target_id = query_data.target_id();
-  HARD_ASSERT(target_ids.find(target_id) == target_ids.end(),
-              "Tried to allocate an already allocated query: %s",
-              query.ToString());
-  target_ids[target_id] = query_data;
+  HARD_ASSERT(
+      query_data_by_target_.find(target_id) == query_data_by_target_.end(),
+      "Tried to allocate an already allocated query: %s", query.ToString());
+  query_data_by_target_[target_id] = query_data;
+  target_id_by_query_[query] = target_id;
   return query_data;
 }
 
 void LocalStore::ReleaseQuery(const Query& query) {
   persistence_->Run("Release query", [&] {
-    absl::optional<QueryData> query_data = query_cache_->GetTarget(query);
+    absl::optional<QueryData> query_data = GetQueryData(query);
```

Comment:
Port of https://github.com/firebase/firebase-android-sdk/blob/d8ff6a54d12fd676a31c2c331f09fe0a7d033202/firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java#L588, which include the change to not double-save QueryData (https://github.com/firebase/firebase-android-sdk/pull/767)